### ISSUE TITLE
[FEATURE REQUEST]: Modern Addressbook

### ISSUE THREAD

**Issue by Milerius** [KomodoPlatform/atomicDEX-Desktop#534]
**Is your feature request related to a problem? Please describe.**

Not really a problem, but a bad user experience, the redesign of this feature will allow us to be sustainable and have something scalable.

**Describe the solution you'd like**

An address book with 0(1) writing and reading, sortable by name and categorizable.

Example of a valid json data for addressbook:

```json
{
  
  "Milerius": 
  {
    "categories": ["Friend"],
    "address_book": {
      "smart_chain": "RX43484BFZbfQiaaRZXDqskHQhwyCKwBk2",
      "erc_20": "0xA00bF635b2cD52F2b6B4D8cd9B9efd290B97838C",
      "qrc_20": "QiNq5uTbqCfumrJQNirasdCi9hS5hAzRyp",
      "BTC": "a_btc_address"
    }
  },
  
  "Alice":
  {
    "categories": ["Pro", "RH", "Salary Provider"],
    "address_book": {
      "smart_chain": "a_smart_chain_address",
      "BTC": "a_btc_address"
    }
  }
}
```

Picking from this format is very easy:

```cpp
my_transaction transaction; //! a KMD transaction
 for (auto&& cur_contact : my_json_object) {
    if (cur_contact.at("address_book").contains(transaction.coin.type)) {
        //! Send a notification with categories and the contact name, the front end will determine the notification
       //! For example if the contact is Alice and one of the categories is `Salary Provider` then the notification will be something like "Alice send you your salary: Total: 500 KMD (250 €)"
     //! If the contact is Milerius and one of the categories is `Friend` then your notification will be smth like: `You received 500 KMD (250 €) from your friend Milerius`
    }     
 }
```

Requirements:

- Possibility to add one or more categories through tag and nice icon or color for example, we determine a list of pre-existing tags
- Possibility to create your own tags (Global to this specific wallet)
- A valid contact need at least one address in the addressbook and one category (category can be `Standard`)
- Remove the old addressbook logic code
- Data of the addressbook will be stored in `$HOME/.atomic_qt/data/wallet_name.addressbook.json` or `%APPDATA%/atomic_qt/data/wallet_name.addressbook.json` (There utility function in utils.hpp to access those directories)
- Adding/Removing/Updating a contact rewrite the whole addressbook in his location
- A file named qt.addressbook.page.(cpp/hpp) will be created, follow the same rules as other page, a page contains models that handle the addressbook data and utility functions.
- It's possible to pickup a contact from the send functionnality in the wallet page (With a searchbar or a filter to pickup a contact in an easy way)
- It's possible to add an image to a contact (the link to this image can be added in the json) and i suggest it to store in our global data folder (.atomic_qt/data/address_book_avatar) (can follow the same logic as adding a custom coin)
- It's possible to export the whole addressbook json into a single QR-Code to export it to another computer or mobile in the future (With a button share)
- If the coin is a `SmartChain`, `ERC-20` or a `QRC-20` then search by the keyword `smart_chain`, `qrc-20` or `erc-20` otherwise search by ticker name `BTC`
- When you add a coin please check the categories first if it's already exist skip - even better remove it from a proxy list

Usefull links: 

- https://doc.qt.io/qt-5/qabstractlistmodel.html
- https://doc.qt.io/qt-5/qabstracttablemodel.html
- https://github.com/nlohmann/json

Development step:

- We will first hardcode an addressbook with a default json like above check our logic etc and then later add the persistent data logic

**Comment by naezith** [comment]
It would be great if you can provide a visual mockup.

**Comment by Milerius** [comment]
> It would be great if you can provide a visual mockup.

Will try to provide that if i have some times this week, we can inspire from what we have now

**Comment by SylEze** [comment]
The persistent data format of an address book is now formatted as followed:
```json
"addressbook_contacts":
[
    {
        "name": "syl",
        "categories": ["Developer", "C++ lover"],
        "wallets_info":
        [
            {
                "type": "BTC",
                "addresses": { "Wallet of home computer": "some address", "Binance wallet": "another address" }
            },
            {
                "type": "erc-20",
                "addresses": { "My valid erc-20 address": "erc20 address" }
            }
        ]
    },
    {
        "name": "Daxter",
        "categories": ["Friend"],
        "wallets_info":
        [
            {
                "type": "ETH",
                "addresses": { "My eth wallet": "An eth waddress" }
            },
            {
                "type": "BTC",
                "addresses": { "My btc wallet": "A btc address" }
            }
        ]
    }
]
```
```

**Comment by Milerius** [comment]
Implemented in #599
